pastas_plugins/pest/solver.py

```diff
@@ -0,0 +1,1159 @@
+            Control data for the PEST solver. Default is None.
+        pcov : Optional[DataFrame], optional
+            The parameter covariance matrix. Default is None.
+        nfev : Optional[int], optional
+            The number of function evaluations. Default is None.
+        long_names : bool, optional
+            Whether to use long names in the PEST control file. Default is True.
+        **kwargs : dict
+            Additional keyword arguments passed to the BaseSolver.
+
+        Returns
+        -------
+        None
+        """
+        BaseSolver.__init__(self, pcov=pcov, nfev=nfev, **kwargs)
+        # model workspace (for pastas files)
+        self.model_ws = Path(model_ws).resolve()
+        if not self.model_ws.exists():
+            self.model_ws.mkdir(parents=True)
+        # template workspace (for pest files)
+        self.temp_ws = Path(temp_ws).resolve()
+        self.exe_name = Path(exe_name)  # pest executable
+        self.pf = pyemu.utils.PstFrom(
+            original_d=self.model_ws,
+            new_d=self.temp_ws,
+            remove_existing=True,
+            longnames=long_names,
+        )
+        copy_file(self.exe_name, self.temp_ws)  # copy pest executable
+        self.noptmax = noptmax
+        self.control_data = control_data
+        self.run_function = """def run():
+    # load packages
+    from pathlib import Path
+
+    from pandas import read_csv
+    from pastas.io.base import load as load_model
+
+    # base path
+    fpath = Path(__file__).parent
+
+    # load pastas model
+    ml = load_model(fpath / "model.pas")
+
+    # update model parameters
+    parameters = read_csv(fpath / "parameters_sel.csv", index_col=0)
+    for pname, val in parameters.loc[:, "optimal"].items():
+        pname = pname.replace("_g", "_A") if pname.endswith("_g") else pname
+        ml.set_parameter(pname, optimal=val)
+
+    # simulate
+    simulation = ml.simulate()
+    simulation.loc[ml.observations().index].to_csv(fpath / "simulation.csv")"""
+
+    def write_run_function(self):
+        """Write the run function to a file"""
+        with (self.model_ws / "_run_pastas_model.py").open("w") as f:
+            f.write(self.run_function)
+
+    def setup_model(self):
+        """Setup and export Pastas model for PEST optimization"""
+        # setup parameters
+        self.ml.parameters.loc[:, "optimal"] = self.ml.parameters.loc[:, "initial"]
+        self.vary = self.ml.parameters.vary.values.astype(bool)
+        parameters = self.ml.parameters[self.vary].copy()
+        parameters.index = [
+            p.replace("_A", "_g") if p.endswith("_A") else p for p in parameters.index
+        ]
+        parameters.index.name = "parnames"
+        if "constant_d" in parameters.index:
+            if np.isnan(parameters.at["constant_d", "pmin"]):
+                self.ml.set_parameter(
+                    "constant_d", pmin=parameters.at["constant_d", "initial"] - 10.0
```

Comment:
why is constant_d limited to +/- 10 around the initial guess?

Reply:
Because otherwise the values are nan values which pest can't deal with.

Reply:
Why not user a (much) larger bound? 10m is okay in many situations but when simulating a highly dynamic time series, e.g. a system with a huge drawdown, you could theoretically run into these bounds?

Reply:
I'm not sure how pastas does this normally as well? Maybe 2* the standard deviation of the observations?